# Cannanetics
Cannanetics - study of cannabis genetics


The goal of my capstone project is to use machine learning techniques to build predictive models which could utilize the genetic information of cannabis plants, more specifically the variations in its genome, and predict the delta-9-tetrahydrocannabinol (THC) content in the resulting flower from these plants. At the moment the only indicator of the expected THC content of the plants flower is the assumed plant strain, and previous observations of the THC content of the strain. As well the Open Cannabis Project existed at one point with the goal to associate strain names to the actual genetics of the plants, but they never investigated the connection between the plant genetics and THC content. Data science has been used before in bioinformatics to extract information from genetic data, and so I thought it was appropriate to use in this case to determine the relationship of the counts of the different variants in each strain, and the resulting THC content those strains produce in their flower, using supervised learning, to be able to make predictions of the THC content based on the variants in the genome. 
The genetics data from the project mentioned above is still available on Google’s BigQuery. The dataset contains the variants for 1400 cannabis samples, determined by comparing the sample genome to a reference genome, and assigning any variations in the sample a unique identifier based on its location in the genome and type. At a very high level, from this dataset I downloaded the variants which belonged to cannabis genome samples which had a strain name associated with them, mapped each unique variant id to a simpler string identifier, and then collected all the variant IDs for each sample as a text list, resulting in a dataframe where each row contained the unique cannabis sample id, the strain name, number of variants, and the list of variants. 

  The THC content data was collected from the results of a GitHub web scrapper where the lab results for cannabis samples were collected from three lab websites. From this dataset I removed duplicate samples, and I extracted the columns with the name of the sample, and those which contained the THC content, but only for samples of cannabis flower. The THC content information was spread across 3 columns, and to get the total THC content in each sample I summed the contents of those columns. 

  One of the biggest challenges I had to tackle was cleaning up the sample/strain names in the lab data and the strain names in the genetic data, so I could use the names to match the THC content data to the appropriate genetic samples using the common strain names found in both. Therefore I cleaned up and standardized the stain names, and I was the able to match the genetic samples to all applicable lab samples, based on the strain names.  I would have liked to have had the actual THC content of the plants used for the genetic data, but unfortunately this information was unavailable, so I decided to match the genetic samples to as many lab samples as possible, to simulate that even if plants have the same genetics, there are so many factors involved that the THC content wouldn’t be the same for all of them.
  
  This resulted in my final data frame, which contained 5 columns in total, with 4187 rows, for 542 unique cannabis genome samples, and 350 unique strains. The columns contain the unique genetic sample ID (‘sra’) as a string, the strain name (‘strain’) as a string, the variants which exist in the sample as a string list of the variants (‘mapped_variant’), the number of variants in the sample (‘num_variant’) as an integer, and the total THC content (‘total_thc’) as a float, as seen in Figure 1. 

![Final Dataframe](/readmeimages/final_dataset_image.png)
<br>**Figure 1**: Resulting dataframe from data collection, and matching of genetic and lab data

  In terms of EDA I first explored the distribution of the number of variants each sample had, since I wanted to see if I had any outliers, possibly caused by issues in the process of determining the variants in the sample genome, against the reference genome. I also looked at the distribution of the total THC content, to get an idea of the range for prediction, and the distribution of THC contents across my samples. As well I explored the relationship between the total number of variants to the total THC which revealed clustering of the samples. I then performed a train/test split and count vectorized the ‘mapped_variant’ column, treating each list of variants as a document, and all the lists of variants as a corpus. The counts of the variants became my features, and the total THC became the target for the modeling.

  Firstly with the count vectorized data I explored the linear relationship of the count of the variants to the total THC content by determining the correlation between each variant and the total THC content, which showed a majority of the variants did not have a correlation, but there still was enough to justify a linear regression. A Decision Tree Regressor was used to explore if there was a threshold in the amount of certain variants which better determines the total THC, as well as a Random Forest Regressor to continue to explore this threshold relationship while taking advantage of its ability to improve feature importance determination, and reduce the effect of noise in the data, which I expected to have as there are so many variants. I implemented K-neighbors Regressor since I thought the relationship between the count of the variants to the THC content would be much more complex and it would be better to compare strain to strain to predict the THC content. I also decided to use Support Vector Regressor since in my EDA I saw we have some clusters when plotting the THC content against the number of variants, so I thought it may be possible some strains are similar enough to be clustered by their variant counts, and we can define boundaries between them to determine the THC content. 
  
  Evaluation of the models was performed by inspecting the R2 value, it was chosen since it provides a measure of how well observed outcomes are replicated by the model, by giving a measure of the proportion of total variation of outcomes explained by the model. As well I used root mean square error (RMSE), as this parameter indicates the standard deviation of the residuals, or how far the points are from the regression prediction. I was aiming for a R2 value close to 1, and as low of an RMSE as possible.
After the initial modelling was completed, all the models performed pretty equally, explaining ~73% of the variance in the training dataset, and ~60% of the variance in the test dataset, with an RMSE score of ~3% THC for the training dataset, and ~ 4% THC for the test dataset, save for a few of the support vector regressor kernel options. After initial modelling I tuned the hyper parameters of each model, along with the min_df term of the count vectorization. In the hyper parameter tuning stage I first explored the effect of each hyper parameter on the validation scores, and then used GridSearchCV with 5 fold cross validation to determine the best combination of the hyper parameters.
Hyper parameter optimization did not yield the results I expected, I saw very little increase across most models in the R2 value, and very little decrease in the RMSE throughout optimization. Overall, the results of the top 5 models (out of 14) from optimization can be found in Figure 2, with the Random Forest Regressor optimized by just its max_depth performing the best, with an R2 score of 0.653 for the test set, and a test RMSE of 4.61 %THC. I also found that the models scores were generally unaffected by the min_df value until it was increased to 80% of the number of documents.

<img src="/readmeimages/optimized_model_results.png" width="75%"> </img>
**Figure 2**: Top 5 models from optimization, optimization method in brackets



  Even though the R2 of the best model is quite high (higher than I expected) at 0.65, the RMSE is remained quite high, and since we are predicting values between 0-30% THC, having a 95% confidence interval (1.96*RMSE) of ±~10% THC is a poor result, so I feel like I half met my goal, in that I have a model that can predict the THC content of a majority of cannabis plants using their genetics, but not with great confidence. However the result of this project can still be refined and be used to help mitigate the high cost of selective cannabis strain breeding programs. 

  Existing breeding programs are resource expensive, as you have to grow multiple plants to maturity to evaluate their THC content, but with the models built in this project the plants genetics could be used to predict its final THC content, allowing for earlier evaluation in the growing stage, so resource needs can be reduced and focused on the plants of interest. The models in this project can also be extended to predict other characteristics of the plant, including other chemical compounds found in the flower and growth characteristics, which I hope to explore in the future. As well, I did not have a chance to perform PCA, or use TF-IDF vectorization to see if there was an improvement in the models performance, and I plan on exploring their effects in the future.

